### ISSUE TITLE
Issues with your latest APK

### ISSUE THREAD

**Issue by IzzySoft** [tawhidmonowar/WebCapture#35]
The updater at the [IzzyOnDroid repo](https://apt.izzysoft.de/fdroid) just reported multiple issues when pulling in your latest APK:

```
! repo/com.prostudio.urltopdfconverter_3.apk declares sensitive permission(s): android.permission.READ_EXTERNAL_STORAGE android.permission.MANAGE_EXTERNAL_STORAGE
! repo/com.prostudio.urltopdfconverter_3.apk: debuggable or testOnly set in AndroidManifest.xml
! com.prostudio.urltopdfconverter_3.apk (com.tawhid.webcapture) has no metadata!
```

OK, the last one means you've changed the packageName. That means it is an entirely new app, and direct updates to this version are not possible.

The first one means we'd have to add the mentioned permissions (with a short description what they are needed for) to your app's "green list" as otherwise they'd be shown as "warnings".

The one in the middle is the biggest problem here: apps for distribution must have **NEITHER** of the two set. Wondering how those come in, problem number 4 shows up:

```
Signer #1 certificate DN: C=US, O=Android, CN=Android Debug
Signer #1 certificate SHA-256 digest: b411f6432b53509f36236b18bee35e5d81de6b6ec5db4f0061f5ee9fe4875259
Signer #1 certificate SHA-1 digest: e134beef19d6f5ea9d6a6ca6010641ace8c03eb6
Signer #1 certificate MD5 digest: 0e102ad1d91636119597a960d87b49a2
Signer #1 key algorithm: RSA
Signer #1 key size (bits): 2048
```

The APK is signed using a debug key. That's not intended for distribution; please always sign release APKs with a release key!

As for the IzzyOnDroid repo, there's a 5th problem now: APK size increased by almost factor 10 from less than 5 MB to almost 40 MB (per-app size limit here is 30 MB). What happened there? For one, compared with the previous release, I see a load of native library directories, each containing about 5 MB for pdfdium. For 6 architectures. Are there really any Android devices running on MIPS? Even x86 is rather rare (or almost non-existent, if we ignore emulators for a little). 2 armeabi dirs (armeabi, armeabi-v7a)? Any devices for the former? All those `*.so` files are stored uncompressed inside the APK, so each dropped unnecessary ABI would mean 5+ MB saved. An APK with just the 2 ARM ABIs would probably be around 15 MB. Per-ABI APKs most likely around 10 MB each or less.

Can we get those issues solved? Thanks in advance!

**Comment by tawhidmonowar** [comment]
Thank you, I will try to fix them as soon as possible.

**Comment by tawhidmonowar** [comment]
**1.** I understand that changing the package name creates a new app. Given that, we should proceed with publishing this version as a new app. This will ensure that users can access the latest features and improvements.

**Comment by IzzySoft** [comment]
> About those permissions

Thanks for the details! As the minimal Android version required for your app is Android 6 (sdkVersion 23), I'd have thought you'd use SAF (Storage Access Framework) instead. The integrated PDFReader to open any PDF files on the device of course makes it reasonable to request full access.

> we should proceed with publishing this version as a new app

If that's intended, that part of course is fine. But the crucial phe point is the debug key there – which is a show-stopper ([see here](https://android.izzysoft.de/articles/named/iod-scan-apkchecks?lang=en#flags) for some background). And the APK size: with 36 MB, it's beyond the per-app limit at IzzyOnDroid (which is 30 MB). Hence my suggestion for split APKs.

So without those two issues fixed, we cannot proceed with publishing, sorry.

**Comment by IzzySoft** [comment]
PS: shouldn't the change of the package name also be mentioned in the "what's changed" section on the release notes? :wink:

**Comment by tawhidmonowar** [comment]
Hi, I've now resolved both concerns.

1. The APK has been signed with the release key.
2. I split the APKs.

Let me know if there anything else need to fix ☺️
Best regards,
Tawhid Monowar

**Comment by IzzySoft** [comment]
Thanks! Integrating that now. One finding by the scanners here:

```
SigningBlock blobs:
-------------------
0x504b4453 (DEPENDENCY_INFO_BLOCK; GOOGLE)
```

This can easily be avoided with a minor addition to your `build.gradle`:

```gradle
android {
    dependenciesInfo {
        // Disables dependency metadata when building APKs.
        includeInApk = false
        // Disables dependency metadata when building Android App Bundles.
        includeInBundle = false
    }
}
```

For some background: that BLOB is supposed to be just a binary representation of your app's dependency tree. But as it's encrypted with a public key belonging to Google, only Google can read it – and nobody else can even verify what it really contains. More details can be found e.g. here: [Ramping up security: additional APK checks are in place with the IzzyOnDroid repo](https://android.izzysoft.de/articles/named/iod-scan-apkchecks#blobs).

----

At IzzyOnDroid we support [Reproducible Builds](https://reproducible-builds.org/) (see: [Reproducible Builds, special client support and more at IzzyOnDroid](https://android.izzysoft.de/articles/named/iod-rbs-mirrors-clients)). Trying for yours, I was able to successfully generate the APK using `./gradlew assembleRelease`, but the resulting APKs were not identical: the arm64 build here ended up with pdfium native libs for ALL architectures:

```diff
  -rw-r--r--  0.0 unx       56 b-       52 defN 1981-01-01 01:01:02 37fbe4f4 META-INF/com/android/build/gradle/app-metadata.properties
- -rw-r--r--  0.0 unx     1720 b-     1720 stor 1981-01-01 01:01:02 30f4a7ff assets/dexopt/baseline.prof
+ -rw-r--r--  0.0 unx     1721 b-     1721 stor 1981-01-01 01:01:02 76417266 assets/dexopt/baseline.prof
  -rw-r--r--  0.0 unx      183 b-      183 stor 1981-01-01 01:01:02 344218f4 assets/dexopt/baseline.profm
- -rw-r--r--  0.0 unx  8971488 b-  3363363 defN 1981-01-01 01:01:02 a7a6afba classes.dex
+ -rw-r--r--  0.0 unx  8971492 b-  3363388 defN 1981-01-01 01:01:02 85b97d77 classes.dex
  -rw-r--r--  0.0 unx     1532 b-      634 defN 1981-01-01 01:01:02 64b746bb classes2.dex
  -rw-r--r--  0.0 unx   352068 b-   144775 defN 1981-01-01 01:01:02 af8d0ddb classes3.dex
  -rw-r--r--  0.0 unx   493432 b-   493432 stor 1981-01-01 01:01:02 e13554fa lib/arm64-v8a/libjniPdfium.so
  -rw-r--r--  0.0 unx  4818288 b-  4818288 stor 1981-01-01 01:01:02 a84f166d lib/arm64-v8a/libmodpdfium.so
+ -rw-r--r--  0.0 unx   107792 b-   107792 stor 1981-01-01 01:01:02 d4d02603 lib/armeabi-v7a/libjniPdfium.so
+ -rw-r--r--  0.0 unx  4466016 b-  4466016 stor 1981-01-01 01:01:02 13f7d27a lib/armeabi-v7a/libmodpdfium.so
+ -rw-r--r--  0.0 unx   111888 b-   111888 stor 1981-01-01 01:01:02 e4cf5620 lib/armeabi/libjniPdfium.so
+ -rw-r--r--  0.0 unx  4466016 b-  4466016 stor 1981-01-01 01:01:02 13f7d27a lib/armeabi/libmodpdfium.so
+ -rw-r--r--  0.0 unx   230088 b-   230088 stor 1981-01-01 01:01:02 c2ad38a2 lib/mips/libjniPdfium.so
+ -rw-r--r--  0.0 unx  5395812 b-  5395812 stor 1981-01-01 01:01:02 ddc460ae lib/mips/libmodpdfium.so
+ -rw-r--r--  0.0 unx   181500 b-   181500 stor 1981-01-01 01:01:02 c3e3baa4 lib/x86/libjniPdfium.so
+ -rw-r--r--  0.0 unx  5506764 b-  5506764 stor 1981-01-01 01:01:02 f127bffa lib/x86/libmodpdfium.so
+ -rw-r--r--  0.0 unx   493736 b-   493736 stor 1981-01-01 01:01:02 a2f5113d lib/x86_64/libjniPdfium.so
+ -rw-r--r--  0.0 unx  5494880 b-  5494880 stor 1981-01-01 01:01:02 d38862ad lib/x86_64/libmodpdfium.so
  -rw----     2.4 fat     1719 b-      777 defN 1981-01-01 01:01:02 e5f78a88 DebugProbesKt.bin
  -rw----     2.4 fat        6 b-        6 stor 1981-01-01 01:01:02 d3bb691b META-INF/androidx.activity_activity.version
  -rw----     2.4 fat        6 b-        6 stor 1981-01-01 01:01:02 45c2cf0d META-INF/androidx.annotation_annotation-experimental.version
@@ -74,7 +84,7 @@
  -rw----     2.4 fat     9256 b-     2938 defN 1981-01-01 01:01:02 44282074 org/apache/commons/validator/resources/validator_1_2_0.dtd
  -rw----     2.4 fat     9256 b-     2944 defN 1981-01-01 01:01:02 efef17c7 org/apache/commons/validator/resources/validator_1_3_0.dtd
  -rw----     2.4 fat     9488 b-     2991 defN 1981-01-01 01:01:02 a89b4e13 org/apache/commons/validator/resources/validator_1_4_0.dtd
- -rw----     0.0 fat     5944 b-     1709 defN 1981-01-01 01:01:02 8fb536a5 AndroidManifest.xml
+ -rw----     0.0 fat     5944 b-     1707 defN 1981-01-01 01:01:02 65dfe20a AndroidManifest.xml
  -rw----     0.0 fat      744 b-      343 defN 1981-01-01 01:01:02 b5e9f5d2 res/-1.xml
```

Further, `versionCode` differs in `AndroidManifest.xml`:

```diff
   E: manifest (line=2)
-    A: http://schemas.android.com/apk/res/android:versionCode(0x0101021b)=3003
+    A: http://schemas.android.com/apk/res/android:versionCode(0x0101021b)=3
     A: http://schemas.android.com/apk/res/android:versionName(0x0101021c)="1.0.2" (Raw: "1.0.2")
```

Also reflected by `classes.dex`:

```diff
       access        : 0x0019 (PUBLIC STATIC FINAL)
-      value         : 6003
+      value         : 3
     #4              : (in Lcom/tawhid/webcapture/BuildConfig;)
```

This might have been caused by my disabling the other ABIs using `sed` on your `build.gradle`:

    sed -r '/abi \{/,/}/ { s/include.*/include "arm64-v8a"/ ; s/universalApk true/universalApk false/ }' -i app/build.gradle

But there was no other way avoiding to build all ABIs when I only needed one (resource wastage). I've tried to just keep in 2 ABIs, same result. Kept it untouched, same result. It seems it always builds only the full APK. Here's the output from the end of the build with the `build.gradle` left untouched:

```
BUILD SUCCESSFUL in 1m 38s
38 actionable tasks: 38 executed
+ find . -name '*.apk'
./app/build/outputs/apk/release/app-release-unsigned.apk
```

Just a single APK where I'd have expected 6. Taking a closer look: You didn't build from the commit the tag points to, did you? Because the `build.gradle` there has no splits defined.

OK, trying from d649b34384e2b93b13e309e070c83885fd48c3be then – the commit you've introduced the splits with. Better now; the additional libs are gone. But it's still not RB:

```diff
  -rw-r--r--  0.0 unx       56 b-       52 defN 1981-01-01 01:01:02 37fbe4f4 META-INF/com/android/build/gradle/app-metadata.properties
- -rw-r--r--  0.0 unx     1720 b-     1720 stor 1981-01-01 01:01:02 30f4a7ff assets/dexopt/baseline.prof
+ -rw-r--r--  0.0 unx     1720 b-     1720 stor 1981-01-01 01:01:02 40e8ea66 assets/dexopt/baseline.prof
  -rw-r--r--  0.0 unx      183 b-      183 stor 1981-01-01 01:01:02 344218f4 assets/dexopt/baseline.profm
- -rw-r--r--  0.0 unx  8971488 b-  3363363 defN 1981-01-01 01:01:02 a7a6afba classes.dex
+ -rw-r--r--  0.0 unx  8971496 b-  3363310 defN 1981-01-01 01:01:02 26eb0ceb classes.dex
  -rw-r--r--  0.0 unx     1532 b-      634 defN 1981-01-01 01:01:02 64b746bb classes2.dex
```

Looking at the Dex diff, it looks like the `versionCode` problem again when trying to build a single ABI:

```diff
       access        : 0x0019 (PUBLIC STATIC FINAL)
-      value         : 6003
+      value         : 3003
     #4              : (in Lcom/tawhid/webcapture/BuildConfig;)
```

But checking both APKs directly, `versionCode` matches for them: they both show `3003`. So this must be something else. Last try: building from eb1b1dd503595e7aaa8935933591f53f5f4ea250 where you applied another minor change. No dice, same issue.

Funnily, when I build from that commit leaving all ABIs enabled, it is RB. I'll use it like that for now then – though that's a bit wasting resources…

We'd appreciate if you could help keeping your build reproducible. We've prepared some [hints on reproducible builds](https://gitlab.com/IzzyOnDroid/repo/-/wikis/Reproducible-Builds) for that. The first basic rule is: always build the APK from a clean tree at the commit the tag points to :wink: (please remember that for **future** releases – do not change the current ones as they've already been distributed now)

Thanks a lot!

cc @obfusk

**Comment by tawhidmonowar** [comment]
@IzzySoft is this app live on F-Droid?

**Comment by tawhidmonowar** [comment]
@licaon-kter can you summarize the issues? 🙂

**Comment by IzzySoft** [comment]
@tawhidmonowar your app is [live at IzzyOnDroid with the new appId](https://apt.izzysoft.de/packages/com.tawhid.webcapture) and shows as [Reproducible Build](https://reproducible-builds.org/) – so be welcome to [pick a badge](https://gitlab.com/IzzyOnDroid/repo/#if-my-app-is-available-here-do-you-have-a-badge-i-can-use-to-link-to-it) to link there e.g. from your Readme :smiley: The only open issue from our end (at IzzyOnDroid) is the signing block blob, which still needs fixing ([see here](https://github.com/tawhidmonowar/WebCapture/issues/35#issuecomment-2328186949) for details):

![image](https://github.com/user-attachments/assets/08bd46b2-33b2-42ee-9e32-d46e0dd28d4b)

@licaon-kter any other issues you see amiss for F-Droid?

**Comment by licaon-kter** [comment]
@IzzySoft since it will be a new app, we will pursue `0x504b4453 (DEPENDENCY_INFO_BLOCK; GOOGLE)` being removed

@tawhidmonowar that ^^^ aside, your "new" PDF stuff uses a lib from 8(!!!) years ago?  https://github.com/tawhidmonowar/WebCapture/blob/1.0.2/app/build.gradle#L46

which depends on https://github.com/DImuthuUpe/AndroidPdfViewer/blob/2.3.0/android-pdf-viewer/build.gradle#L41

which pulls random binaries like https://github.com/barteksc/PdfiumAndroid/tree/pdfium-android-1.5.0/src/main/jni/lib 

fyi we disabled a bunch of apps using a newer version of that lib 2(!!) years ago: https://gitlab.com/fdroid/fdroiddata/-/merge_requests/11496 per ref: https://divestos.org/misc/appsec.txt
